Introduce new 'docker run' validation command (#3716)
Kitware/CDash: docs/xml-validation.md

Reviewer: I think we should give this a more descriptive name.  Perhaps `xml-validation.md`?

Author: I've updated the name of the file. Thanks!

# Validating XML files for CDash

CDash maintains a set of schema files for the expected structure of
the incoming XML files that will be parsed.  Normally, this is not a
problem as CTest writes properly formatted files.

In the rare case that the XML files are not being created by CTest, we have a
few opportunities for the user to be informed that there are problems
with the XML files that are uploaded

1. Prior to Submission

The CDash Docker container has an entry point which will look for files within
a directory which has been mounted to the container.  The directory of files
should be mounted to `/tmp/validate` using the `-v` flag.  The entry point
to call `validate-xml` should be executed via the `docker run` command.

```bash
docker run -v /example/host/directory/xml:/tmp/validate kitware/cdash validate-xml
```

> For example, when all files are properly validated, the report would print to
> the screen like this:

Reviewer: ```suggestion
When all files are properly validated, the report looks like this:
```

```bash
Validated file: /tmp/validate/Build.xml.
Validated file: /tmp/validate/Configure.xml.
Validated file: /tmp/validate/Coverage.xml.
Validated file: /tmp/validate/CoverageLog-0.xml.
Validated file: /tmp/validate/Done.xml.
Validated file: /tmp/validate/Test.xml.
```

> Errors in a file will be printed to the screen:

Reviewer: ```suggestion
Errors in a file will be printed to `stderr`:
```

```bash
WARNING: Element 'Site': The attribute 'Name' is required but missing.
 in /tmp/validate/Build.xml, line: 25, column: 0
Validated file: /tmp/validate/Configure.xml.
Validated file: /tmp/validate/Coverage.xml.
Validated file: /tmp/validate/CoverageLog-0.xml.
FAILED: Some XML file checks did not pass!
```

2. During Submission

For validation during the submission process, CDash will look at the
`VALIDATE_SUBMISSIONS` environment variable.  The value of the variable
determines how strict CDash will be when accepting XML files that may be
invalid.

* `SILENT`
  * CDash will log a message to its log file about the failures it finds,
    but will continue to process the content of the file.
  * This is the default value.
* `WARN`
  * CDash will log a message to its log file about the failures it finds.
    It will also send the failure results in the `message` attribute of the
    response from the HTTP request but will continue processing the content
    of the file.
* `REJECT`

>   * CDash will log a message to its log file about the failures it finds.
>     It will also halt processing the file, returning a 400 response code, and
>     informing the user the failure results

Reviewer: ```suggestion
  * CDash will log a message to its log file about the failures it finds.
    It will also halt processing the file, returning a 400 response code, and
    informing the user the failure results.
```

# Validating XML files for CDash

CDash maintains a set of schema files for the expected structure of
the incoming XML files that will be parsed.  Normally, this is not a
problem as CTest writes properly formatted files.

In the rare case that the XML files are not being created by CTest, we have a
few opportunities for the user to be informed that there are problems
with the XML files that are uploaded

1. Prior to Submission

The CDash Docker container has an entry point which will look for files within
a directory which has been mounted to the container.  The directory of files
should be mounted to `/tmp/validate`.  This entry point called `validate-xml`
should be executed via the `run` command.

For example:
```bash

> CDash$ sudo docker run -v <full path>/20260504-1334:/tmp/validate kitware/cdash validate-xml

Reviewer: 1. `sudo` shouldn't be necessary on most systems, right?
2. Can you put an example placeholder like `/example/host/directory` instead of `<full path>`?  The `20260504-1334` component is also unnecessary, since many users likely put these files in an arbitrary filesystem directory.
3. I think we should better distinguish between the command to be copied and pasted and output.  That way, users can more easily select the command.

Author: I think I got them all, please look again!
